### PR TITLE
Init utils4 test

### PR DESCRIPTION
**Pull request by ductm54**

### CODE REVIEW

File: test/sol5/utils4.js [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+        rate = MAX_RATE
+        srcQty = MAX_QTY.add(new BN(1))
+        await expectRevert(
+            utils4.mockCalcDstQty(srcQty, srcDecimal, dstDecimal, rate),
+            "srcQty > MAX_QTY"
+        )
+    });
+
+
+    it("check src qty calculation for high quantities.", async function () {
```
Comment (ilanDoron) [inline test/sol5/utils4.js]: please split into more tests.
per test need to test 1 thing.
so can split for srcDecimal > destDecimal.
destDecimals > srcDecimals
check revert

File: test/sol5/utils4.js [hunk 2]
```diff
@@ -0,0 +1,320 @@
+const MockUtils4 = artifacts.require("MockUtils4.sol")
+const Helper = require("../helper.js");
+const TestToken = artifacts.require("Token.sol");
+const TokenNoDecimal = artifacts.require("TokenNoDecimal.sol");
+const { expectRevert } = require('@openzeppelin/test-helpers');
+const BN = web3.utils.BN;
+const { BPS, precisionUnits, ethDecimals, ethAddress, zeroAddress, emptyHint } = require("../helper.js");
+
+
+const PRECISION = new BN(10).pow(new BN(18));
```
[the remark this answers is not in the document]
Comment (ductm54) [reply test/sol5/utils4.js]: update

File: test/sol5/utils4.js [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+const MAX_QTY = new BN(10).pow(new BN(28));
+const MAX_RATE = new BN(10).pow(new BN(25));
+
```
[the remark this answers is not in the document]
Comment (ductm54) [reply test/sol5/utils4.js]: update

File: test/sol5/utils4.js [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        let tokenBalance = 600;
+        let user = accounts[8]
+
+        await token.transfer(user, tokenBalance);
+
+        let balance = await utils4.mockGetBalance(token.address, user);
+        Helper.assertEqual(balance, tokenBalance);
+    })
+
+    it("should check get decimals for normal token", async function () {
```
[the remark this answers is not in the document]
Comment (ductm54) [reply test/sol5/utils4.js]: break into smaller test

File: test/sol5/utils4.js [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+        let token = await TestToken.new("regular", "reg", 16);
+        await utils4.mockSetDecimals(token.address);
+        Helper.assertEqual(await utils4.mockGetDecimals(token.address), 16);
+    });
+
+    it("should check get update decimals for normal token", async function () {
+        // check logical
+        Helper.assertEqual(await utils4.mockCheckGetUpdateDecimals.call(ethAddress), ethDecimals)
+        let token = await TestToken.new("regular", "reg", 16);
+        Helper.assertEqual(await utils4.mockCheckGetUpdateDecimals.call(token.address), 16);
```
[the remark this answers is not in the document]
Comment (ductm54) [reply test/sol5/utils4.js]: add mock read

File: test/sol5/utils4.js [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+            "src - dst > MAX_DECIMALS",
+        )
+    });
+
+    it("test min of", async function () {
+        Helper.assertEqual(3, await utils4.mockMinOf(3, 10))
+        Helper.assertEqual(3, await utils4.mockMinOf(10, 3))
+    });
+
+})
```
[the remark this answers is not in the document]
Comment (ductm54) [reply test/sol5/utils4.js]: added

File: test/sol5/utils4.js [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+        let expectedDestQty = Helper.calcDstQty(srcQty, srcDecimal, dstDecimal, rate);
+        let reportedDstQty = await utils4.mockCalcDstQty(srcQty, srcDecimal, dstDecimal, rate);
+        Helper.assertEqual(expectedDestQty, reportedDstQty, "unexpected dst qty");
+        Helper.assertEqual(
+            expectedDestQty,
+            await utils4.mockCalcDestAmount(srcAddress, dstAddress, srcQty, rate),
+            "unexpected destamount"
+        );
+
+        //next check when dest decimals > src decimals
```
[the remark this answers is not in the document]
Comment (ductm54) [reply test/sol5/utils4.js]: fix

File: test/sol5/utils4.js [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+    it("should check get decimals for token without decimals API. see reverts", async function () {
+        let tokenNoDecimal = await TokenNoDecimal.new("noDec", "dec", 18);
+
+        //now get deicmals to see values
+        await expectRevert.unspecified(utils4.mockSetDecimals(tokenNoDecimal.address));
+        await expectRevert.unspecified(utils4.mockGetDecimals(tokenNoDecimal.address));
+        await expectRevert.unspecified(utils4.mockCheckGetUpdateDecimals(tokenNoDecimal.address));
+    });
+
+    it("check dest qty calculation for high quantities.", async function () {
```
[the remark this answers is not in the document]
Comment (ductm54) [reply test/sol5/utils4.js]: break into smaller test

File: test/sol5/utils4.js (hunk 1 above)
Comment (ductm54) [reply test/sol5/utils4.js]: break into smaller test

**Pull request merged**